File: CS/SLApp/MainPageViewModel.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.ComponentModel;
using DevExpress.Xpf.Printing;
using DevExpress.Xpf.Core.Commands;
using System.Linq;

namespace SLApp
{
    public class MainPageViewModel : INotifyPropertyChanged
    {
        private ReportPreviewModel _previewModel;
        public ReportPreviewModel PreviewModel
        {
            get
            {
                return _previewModel;
            }
        }

        private ReportPreviewModel _popupPreviewModel;
        public ReportPreviewModel PopupPreviewModel
        {
            get
            {
                return _popupPreviewModel;
            }
        }

        private readonly DelegateCommand<object> _linkClickedCommand;
        public DelegateCommand<object> LinkClickedCommand
        {
            get
            {
                return _linkClickedCommand;
            }
        }

        private readonly DelegateCommand<object> _showProductsCommand;
        public DelegateCommand<object> ShowProductsCommand
        {
            get
            {
                return _showProductsCommand;
            }
        }

        private string _linkText;
        public string LinkText
        {
            get { return _linkText; }
            set
            {
                if (_linkText != value)
                {
                    _linkText = value;
                    OnPropertyChanged("LinkText");
                }
            }
        }


        private int _orderId;
        public int OrderId
        {
            get
            {
                return _orderId;
            }
            set
            {
                if (_orderId != value)
                {
                    _orderId = value;
                    OnPropertyChanged("OrderId");
                }
            }
        }
        /// <summary>
        /// Initializes a new instance of the MainPageViewModel class.
        /// </summary>
        public MainPageViewModel()
        {
            _previewModel = new ReportPreviewModel
            {
                ServiceUri = "../ReportService.svc",
                ReportName = "Reports.OrderReport"
            };

            _popupPreviewModel = new ReportPreviewModel
            {
                ServiceUri = "../ReportService.svc",
                ReportName = "Reports.ProductReport",
            };

            _previewModel.PreviewClick += previewModel_PreviewClick;
            _previewModel.PreviewMouseMove += previewModel_PreviewMouseMove;
            _previewModel.RequestDefaultParameterValues += previewModel_RequestDefaultParameterValues;

            OrderId = 0;
            _linkClickedCommand = new DelegateCommand<object>(LinkClicked);
            _showProductsCommand = new DelegateCommand<object>(ShowProducts);
        }


        protected internal void ShowProducts(object obj)
        {
            var doc = new DocumentPreviewWindow
            {
                Model = _popupPreviewModel,
                Name = "PopupPreview",
            };
            doc.Model.CreateDocument();
            doc.Preview.AutoCreateDocument = true;
            doc.Show();
        }

        protected internal void LinkClicked(object o)
        {
            if (_previewModel.ReportName == "Reports.InvoiceReport")
            {
                _previewModel.ReportName = "Reports.OrderReport";
                _previewModel.CreateDocument();
                
                LinkText = "";
            }
        }

        protected internal void previewModel_RequestDefaultParameterValues(object sender, EventArgs e)
        {
            if (_previewModel.ReportName == "Reports.InvoiceReport")
                _previewModel.Parameters["OrderId"].Value = OrderId;
        }

        protected void previewModel_PreviewMouseMove(object sender, PreviewClickEventArgs e)
        {
            if (e.ElementTag != null && !e.ElementTag.Equals(string.Empty))
                e.Control.Cursor = Cursors.Hand;
        }

        protected void previewModel_PreviewClick(object sender, PreviewClickEventArgs e)
        {
            if (e.ElementTag != null && !e.ElementTag.Equals(string.Empty))
            {
                OrderId = int.Parse(e.ElementTag);
                LinkText = "Go Back";

                _previewModel.ReportName = "Reports.InvoiceReport";
                _previewModel.IsDocumentMapVisible = false;

                if (_previewModel.Parameters.Count > 0)
                    _previewModel.Parameters["OrderId"].Value = OrderId;

                _previewModel.CreateDocument();
            }
        }

        public virtual void OnPropertyChanged(string property)
        {
            PropertyChangedEventHandler handler = PropertyChanged;
            if (handler != null)
                handler(this, new PropertyChangedEventArgs(property));
        }
        public event PropertyChangedEventHandler PropertyChanged;
    }
}

File: VB/SLApp/MainPageViewModel.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Windows.Input
Imports DevExpress.Xpf.Mvvm
Imports DevExpress.Xpf.Printing

Namespace SLApp
	Public Class MainPageViewModel
		Implements INotifyPropertyChanged
		Private privatePreviewModel As ReportPreviewModel
		Public Property PreviewModel() As ReportPreviewModel
			Get
				Return privatePreviewModel
			End Get
			Private Set(ByVal value As ReportPreviewModel)
				privatePreviewModel = value
			End Set
		End Property
		Private privatePopupPreviewModel As ReportPreviewModel
		Public Property PopupPreviewModel() As ReportPreviewModel
			Get
				Return privatePopupPreviewModel
			End Get
			Set(ByVal value As ReportPreviewModel)
				privatePopupPreviewModel = value
			End Set
		End Property
		Private privateLinkClickedCommand As DelegateCommand
		Public Property LinkClickedCommand() As DelegateCommand
			Get
				Return privateLinkClickedCommand
			End Get
			Set(ByVal value As DelegateCommand)
				privateLinkClickedCommand = value
			End Set
		End Property
		Private privateShowProductsCommand As DelegateCommand
		Public Property ShowProductsCommand() As DelegateCommand
			Get
				Return privateShowProductsCommand
			End Get
			Set(ByVal value As DelegateCommand)
				privateShowProductsCommand = value
			End Set
		End Property

		Private _linkText As String
		Public Property LinkText() As String
			Get
				Return _linkText
			End Get
			Set(ByVal value As String)
				If _linkText <> value Then
					_linkText = value
					OnPropertyChanged("LinkText")
				End If
			End Set
		End Property

		''' <summary>
		''' Initializes a new instance of the MainPageViewModel class.
		''' </summary>
		Public Sub New()
			PreviewModel = New ReportPreviewModel With {.ServiceUri = "../ReportService.svc", .ReportName = "Reports.OrderReport"}
			AddHandler PreviewModel.PreviewClick, AddressOf PreviewModel_PreviewClick
			AddHandler PreviewModel.PreviewMouseMove, AddressOf PreviewModel_PreviewMouseMove

			LinkClickedCommand = New DelegateCommand(AddressOf LinkClicked)
			ShowProductsCommand = New DelegateCommand(AddressOf ShowProducts)
		End Sub


		Private Sub ShowProducts()
			Dim model = New ReportPreviewModel With {.ServiceUri = "../ReportService.svc", .ReportName = "Reports.ProductReport"}
			Dim window = New DocumentPreviewWindow With {.Model = model}
			window.Show()
			model.CreateDocument()
		End Sub

		Private Sub LinkClicked()
			If PreviewModel.ReportName <> "Reports.InvoiceReport" Then
				Return
			End If
			PreviewModel.Clear()
			PreviewModel.ReportName = "Reports.OrderReport"
			PreviewModel.CreateDocument()

			LinkText = ""
		End Sub

		Protected Sub PreviewModel_PreviewMouseMove(ByVal sender As Object, ByVal e As PreviewClickEventArgs)
			If (Not String.IsNullOrEmpty(e.ElementTag)) Then
				e.Element.Cursor = Cursors.Hand
			End If
		End Sub

		Protected Sub PreviewModel_PreviewClick(ByVal sender As Object, ByVal e As PreviewClickEventArgs)
			Dim orderId As Integer
			If (Not Integer.TryParse(e.ElementTag, orderId)) Then
				Return
			End If

			LinkText = "Go Back"

			PreviewModel.ReportName = "Reports.InvoiceReport"
			PreviewModel.IsDocumentMapVisible = False

			PreviewModel.Clear()
			PreviewModel.Parameters("OrderId").Value = orderId

			PreviewModel.CreateDocument()
		End Sub

		Public Sub OnPropertyChanged(ByVal [property] As String)
			RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs([property]))
		End Sub

		Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
	End Class
End Namespace
